CI: NVHPC for CUDA

ax3l opened the pull request:
## Summary

Add a CI entry that compiles the host C++ side with `nvc++` and the device side with `nvcc`.

We need this compiler pairing to proceed with Perlmutter Nvidia PrgEnv & E4S support.

- [x] first updates to the code base to handle `nvc++` as supported compiler
- [x] triage with Nvidia to find the problem with the type trait containing `__nv_is_extended_device_lambda_closure_type`
- [ ] potentially further code-base fixes/work-arounds for `nvcc` (device) + `nvc++` (host)

## Additional background

This seems to be the default Cray `PrgEnv-nvidia` that is bestowed upon us on Perlmutter.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

ax3l committed:
CI: NVHPC

Add a CI entry that compiles the host C++ side with nvc++ and
the device side with nvcc.

This seems to be the default Cray-ed up `PrgEnv-nvidia` that we
receive on Perlmutter.

ax3l committed:
CMake: NVHPC Support

ax3l committed:
Mark TODOs: PGI -> NVHPC

ax3l commented:
@maxpkatz confirmed the issue and reported it upstream with the NVHPC people to understand the underlying issue. It looks like `nvcc` is not stripping that type trait containing the `__nv_is_extended_device_lambda_closure_type` calls from the code that it is passed to `nvc++`.

ax3l commented:
Type trait problem:
- X-ref: Nvidia issue number 3319881
- marked as resolved for the next release (current release: 21.5)

ax3l commented on .github/workflows/dependencies/dependencies_nvhpc21-7.sh:
```diff
@@ -0,0 +1,29 @@
+wget -q https://developer.download.nvidia.com/hpc-sdk/21.7/nvhpc-21-7_21.7_amd64.deb \
+        https://developer.download.nvidia.com/hpc-sdk/21.7/nvhpc-2021_21.7_amd64.deb
+sudo apt-get update
+sudo apt-get install -y ./nvhpc-21-7_21.7_amd64.deb ./nvhpc-2021_21.7_amd64.deb
```
```
After this operation, 8,983 MB of additional disk space will be used.
```
:eyes:

wyphan replied:
If you wanna go the container route, NVIDIA also provides https://ngc.nvidia.com/catalog/containers/nvidia:nvhpc

ax3l replied:
Thanks! Will use this for the reproducer for Nvidia.
Unfortunately, that container is of the same size or even a bit larger than hand-picking the primary packages we need here.

We should definitely keep it in mind though in case of stability/limits that can be different between docker and apt runs (potentially).

wyphan replied:
There's Shifter on Cori to deal with Docker containers... https://docs.nersc.gov/development/shifter/how-to-use/
And there is the `compile` queue which is free-of-charge and intended for CI. https://software.nersc.gov/NERSC/nersc-ci-docs/-/wikis/Home#queues-and-queue-policies (requires NERSC login)

ax3l commented:
@maxpkatz @WeiqunZhang @kngott we are stuck on the following problems even with NVHPC 21.7:

### Type Trait

```
"/home/runner/work/amrex/amrex/Src/Base/AMReX_TypeTraits.H", line 86: error: type name is not allowed
  struct MaybeHostDeviceRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::false_type { 
                                                                                                   ^
```

### Launch

Commenting out the above type trait, we get stuck on some wrapping logic:
```
/opt/nvidia/hpc_sdk/Linux_x86_64/21.7/compilers/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/opt/nvidia/hpc_sdk/Linux_x86_64/21.7/compilers/bin/nvc++  -I/home/axel/src/amrex/Src/Base -I/home/axel/src/amrex/Src/Base/Parser -I/home/axel/src/amrex/Src/Boundary -I/home/axel/src/amrex/Src/AmrCore -I/home/axel/src/amrex/Src/Amr -I/home/axel/src/amrex/Src/LinearSolvers/MLMG -I/home/axel/src/amrex/Src/LinearSolvers/Projections -I/home/axel/src/amrex/Src/Particle -I/home/axel/src/amrex/build -isystem=/opt/nvidia/hpc_sdk/Linux_x86_64/21.7/comm_libs/openmpi/openmpi-3.1.5/include -O3 -DNDEBUG --generate-code=arch=compute_80,code=[compute_80,sm_80] --expt-relaxed-constexpr --expt-extended-lambda -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored -maxrregcount=255 -Xcudafe --display_error_number --Werror cross-execution-space-call --use_fast_math -std=c++17 -MD -MT Src/CMakeFiles/amrex.dir/Base/AMReX_BlockMutex.cpp.o -MF CMakeFiles/amrex.dir/Base/AMReX_BlockMutex.cpp.o.d -x cu -dc /home/axel/src/amrex/Src/Base/AMReX_BlockMutex.cpp -o CMakeFiles/amrex.dir/Base/AMReX_BlockMutex.cpp.o

Src/Base/AMReX_BlockMutex.cpp", line 12: error: identifier "__nv_dl_wrapper_t" is undefined
  amrex::launch((N + 255) / 256, 256, Gpu::nullStream(), __nv_dl_wrapper_t< __nv_dl_tag<void (*)(state_t * state, int N) noexcept,(&init_states),1> , typename __nvdl_remove_ref<decltype(N)>::type, typename __nvdl_remove_ref<decltype(state)>::type>( __nv_dl_tag<void (*)(state_t * state, int N) noexcept,(&init_states),1> {} , N, state) 
                                                         ^
```

ax3l commented:
GNUmake reproducer:
```
docker pull nvcr.io/nvidia/nvhpc:21.7-devel-cuda_multi-ubuntu20.04
docker run -it -v $(pwd):/host_pwd -w /host_pwd nvcr.io/nvidia/nvhpc:21.7-devel-cuda_multi-ubuntu20.04

apt update
apt install -y python python3
alias python=python3

make COMP=nvhpc USE_CUDA=TRUE NVCC_CCBIN=nvc++ -f GNUmakefile.in
```
```
Compiling AMReX.cpp ...
nvcc  -ccbin=nvc++ -Xcompiler=' -Werror=return-type -g -O3  -pthread -std=c++14' --std=c++14 -Wno-deprecated-gpu-targets -m64 -arch=compute_70 -code=sm_70 -maxrregcount=255 --expt-relaxed-constexpr --expt-extended-lambda -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored --Werror cross-execution-space-call -lineinfo --ptxas-options=-O3 --ptxas-options=-v --use_fast_math  --forward-unknown-to-host-compiler --Werror ext-lambda-captures-this --display-error-number --diag-error 20092 -x cu -dc   -DAMREX_GIT_VERSION=\"21.08-56-g8783ee957e29\" -DAMREX_USE_CUDA -DAMREX_USE_GPU -DBL_COALESCE_FABS -DAMREX_USE_GPU_RDC -DBL_SPACEDIM=3 -DAMREX_SPACEDIM=3 -DBL_FORT_USE_UNDERSCORE -DAMREX_FORT_USE_UNDERSCORE -DBL_Linux -DAMREX_Linux -DNDEBUG -DAMREX_GPU_MAX_THREADS=256 -Itmp_build_dir/s/3d.nvhpc.CUDA.EXE -I. -I/host_pwd/Src/Base -I/host_pwd/Src/Base/Parser -I/host_pwd/Src/AmrCore -I/host_pwd/Src/Amr -I/host_pwd/Src/Boundary -isystem /usr/local/cuda/include -c /host_pwd/Src/Base/AMReX.cpp -o tmp_build_dir/o/3d.nvhpc.CUDA.EXE/AMReX.o
ptxas info    : 218059 bytes gmem, 72 bytes cmem[3]
ptxas info    : Compiling entry function '_ZN3cub11EmptyKernelIvEEvv' for 'sm_70'
ptxas info    : Function properties for _ZN3cub11EmptyKernelIvEEvv
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 4 registers, 352 bytes cmem[0]
"/host_pwd/Src/Base/AMReX_TypeTraits.H", line 86: error: type name is not allowed
  struct MaybeHostDeviceRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::false_type { 
                                                                                                   ^

"/host_pwd/Src/Base/AMReX_TypeTraits.H", line 86: error: identifier "__nv_is_extended_device_lambda_closure_type" is undefined
  struct MaybeHostDeviceRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::false_type { 
                                                       ^

"/host_pwd/Src/Base/AMReX_TypeTraits.H", line 90: error: type name is not allowed
  struct DefinitelyNotHostRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::true_type { 
                                                                                                     ^

"/host_pwd/Src/Base/AMReX_TypeTraits.H", line 90: error: identifier "__nv_is_extended_device_lambda_closure_type" is undefined
  struct DefinitelyNotHostRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::true_type { 
                                                         ^

4 errors detected in the compilation of "/tmp/tmpxft_000014a2_00000000-6_AMReX.cudafe1.cpp".
make: *** [/host_pwd/Tools/GNUMake/Make.rules:200: tmp_build_dir/o/3d.nvhpc.CUDA.EXE/AMReX.o] Error 2
```

ax3l commented:
Filed as nvidia developers bug no. 3366114

ax3l committed:
nvcc -ccbin=nvc++: avoid unsupported traits

Work-around NV bug 3366114
```
"/home/runner/work/amrex/amrex/Src/Base/AMReX_TypeTraits.H", line 86: error: type name is not allowed
  struct MaybeHostDeviceRunnable< T, std::enable_if_t< __nv_is_extended_device_lambda_closure_type(T)> >  : public std::false_type {
                                                                                                   ^
```

ax3l committed:
NVHPC: 21.7 -> 21.9

ax3l commented:
Jippey, progress with NVHPC 21.9.

New Fortran issue: probably just a way to hint the compiler where to create module files:
```
[ 30%] Building Fortran object Src/CMakeFiles/amrex.dir/Base/AMReX_FILCC_3D.F90.o
cd /home/runner/work/amrex/amrex/build/Src && /opt/nvidia/hpc_sdk/Linux_x86_64/21.9/compilers/bin/nvfortran -DAMREX_LANG_FORT -DBL_LANG_FORT -I/home/runner/work/amrex/amrex/Src/Base -I/home/runner/work/amrex/amrex/Src/Base/Parser -I/home/runner/work/amrex/amrex/Src/Boundary -I/home/runner/work/amrex/amrex/Src/AmrCore -I/home/runner/work/amrex/amrex/Src/Amr -I/home/runner/work/amrex/amrex/Src/LinearSolvers/MLMG -I/home/runner/work/amrex/amrex/Src/LinearSolvers/Projections -I/home/runner/work/amrex/amrex/Src/Particle -I/home/runner/work/amrex/amrex/build -I/opt/nvidia/hpc_sdk/Linux_x86_64/21.9/comm_libs/openmpi/openmpi-3.1.5/include -fast -O3 -module ../mod_files -MD -MT Src/CMakeFiles/amrex.dir/Base/AMReX_FILCC_3D.F90.o -MF CMakeFiles/amrex.dir/Base/AMReX_FILCC_3D.F90.o.d -c /home/runner/work/amrex/amrex/Src/Base/AMReX_FILCC_3D.F90 -o CMakeFiles/amrex.dir/Base/AMReX_FILCC_3D.F90.o
NVFORTRAN-F-0004-Unable to open MODULE file amrex_fort_module.mod (/home/runner/work/amrex/amrex/Src/Base/AMReX_FILCC_3D.F90: 25)
NVFORTRAN/x86-64 Linux 21.9-0: compilation aborted
```

ax3l committed:
NVHPC: Disable Fortran for now

maximumcats commented:
For reference, here is what the Fortran compile line looks like in GNU Make, where the module directory is being pointed to the tmp_build_dir where all the other temporaries go:
```
nvfortran  -O2 -fast -Mnomain -gpu=cc70,fastmath -Mcuda=lineinfo CUDA_HOME= -gpu=maxregcount:255 -Mdclchk  -mp=gpu -Minfo=mp -gpu=cc70  -module tmp_build_dir/o/3d.nvhpc.OMP_OFFLOAD.CUDA.EXE -Itmp_build_dir/o/3d.nvhpc.OMP_OFFLOAD.CUDA.EXE -DBL_LANG_FORT -DAMREX_LANG_FORT -DCASTRO_DEVICE= -DCASTRO_FORT_DEVICE= -DCRSEGRNDOMP -DAMREX_GIT_VERSION=\"21.09-45-g9256409e2aae\" -DAMREX_USE_OMP_OFFLOAD -DAMREX_USE_CUDA -DAMREX_USE_GPU -DBL_COALESCE_FABS -DAMREX_USE_GPU_RDC -DBL_SPACEDIM=3 -DAMREX_SPACEDIM=3 -DBL_FORT_USE_UNDERSCORE -DAMREX_FORT_USE_UNDERSCORE -DBL_Linux -DAMREX_Linux -DNDEBUG -DAMREX_GPUS_PER_SOCKET=4 -DAMREX_GPUS_PER_NODE=8 -DAMREX_GPU_MAX_THREADS=256 -Itmp_build_dir/s/3d.nvhpc.OMP_OFFLOAD.CUDA.EXE -I. -I../amrex/Src/Base -I../amrex/Src/Base/Parser -I../amrex/Src/AmrCore -I../amrex/Src/Amr -I../amrex/Src/Boundary -I../Source  -c ../Source/riemann.F90 -o tmp_build_dir/o/3d.nvhpc.OMP_OFFLOAD.CUDA.EXE/riemann.o
```

ax3l committed:
Remove Work-Around: CUDA Type Trait

ax3l commented:
@maxpkatz @WeiqunZhang ready for merge.

We can address Fortran NVHPC in a follow-up.

ax3l commented on Src/Base/AMReX_Extension.H:
```diff
@@ -61,6 +63,8 @@
 #elif defined(__PGI)
 #define AMREX_PRAGMA_SIMD _Pragma("loop ivdep")
 
+//__NVCOMPILER
```
@maxpkatz do you know the pragma SIMD equivalent for NVHPC?

ax3l replied:
Doesn't matter for now:
- `__PGI` is still defined, so `_Pragma("loop ivdep")` it is :)
```suggestion
#elif defined(__NVCOMPILER)
#define AMREX_PRAGMA_SIMD _Pragma("loop ivdep")
```

ax3l commented on Src/Base/AMReX_Extension.H:
```diff
@@ -94,6 +98,8 @@
 #elif defined(__PGI)
 #define AMREX_FORCE_INLINE inline
 
+//__NVCOMPILER
```
Maybe I should try:
```suggestion
#elif defined(__NVCOMPILER)
#define AMREX_FORCE_INLINE inline
```

ax3l committed:
Try if NVHPC understands `inline`

ax3l committed:
Add NVC++ SIMD

Same as PGI

ax3l committed:
NVC++: Also no omp threadprivate

Just as PGI predecessor:
```
NVC++-S-0155-threadprivate class variables require constructors   (/opt/nvidia/hpc_sdk/Linux_x86_64/21.9/comm_libs/openmpi/openmpi-3.1.5/include/openmpi/ompi/mpi/cxx/functions_inln.h: 144)
NVC++/x86-64 Linux 21.9-0: compilation completed with severe errors
```

ax3l commented:
@maxpkatz @WeiqunZhang if we merge this, we can already try to get WarpX 21.10 ready for E4S Perlmutter builds.

WeiqunZhang commented:
nvhpc seems to be much slower than nvcc.

maximumcats commented:
I suspect the cost is primarily getting the large .deb files and installing them (this is typically fairly slow since they are multi-GB installers with lots of files to copy).

wyphan commented:
Which is why I suggested using Docker/Shifter. I can try this once NVIDIA updates their official Docker images to 21.9.

WeiqunZhang commented:
It's both. Although the "Dependencies" step was 5m 35s, much slower than 1m 17s for cuda, the "Build & Install" step was 36m 48s, slower than CUDA's 29m 36s.

ax3l commented:
@wyphan thanks, I tested this and the container is even slightly larger than the `.deb`:
https://github.com/AMReX-Codes/amrex/pull/2066#discussion_r695178725
> Unfortunately, that container is of the same size or even a bit larger than hand-picking the primary packages we need here.

Also the NGC containers are a bit delayed these days, so there is no official 21.9 yet, even if it were smaller and thus faster to download/partition.

wyphan commented:
@ax3l The idea is caching the container on Cori/Perlmutter filesystem using Shifter so we only download it once per release, instead of every time the CI pipeline is run.

ax3l commented:
Oh, but we want to test NVHPC in PRs as well, otherwise we need to catch up once a release or so.
We tried this historically (post-merge CI) and it's not very sustainable with enough contributors, I fear
